File: 卡码/代码随想录/Python/动态规划/02-背包问题/03-多重背包/Kama_0056.py
#56. 携带矿石资源
from typing import *

def main():
  [c,n]=map(int,input().split())
  w:List[int]=list(map(int,input().split()))
  v:List[int]=list(map(int,input().split()))
  k:List[int]=list(map(int,input().split()))
  dp:List[int]=[0]*(c+1)
  for i in range(n):
    for j in range(c,w[i]-1,-1):
      for l in range(1,k[i]+1):
        if j<l*w[i]:
          break
        dp[j]=max(dp[j],dp[j-l*w[i]]+l*v[i])
  print(dp[c])

if __name__=="__main__":
  main()

File: 卡码/代码随想录/JS/动态规划/02-背包问题/03-多重背包/Kama_0056.js
//56. 携带矿石资源
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

(async function(){
	const [c,n]=(await readline()).split(" ").map(Number);
	const w=(await readline()).split(" ").map(Number);
	const v=(await readline()).split(" ").map(Number);
	const k=(await readline()).split(" ").map(Number);
	const dp=new Array(c+1).fill(0);
	for(let i=0;i<n;i++){
		for(let j=c;j>=w[i];j--){
			for(let l=1;l<=k[i]&&j>=l*w[i];l++){
				dp[j]=Math.max(dp[j],dp[j-l*w[i]]+l*v[i]);
			}
		}
	}
	console.log(dp[c]);
})();